File: Data Structures & Algorithms/leetcode/arrays-and-hashing/two-integer-sum/submission-3.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> nbsToIndices = new HashMap<>();
        int[] output = new int[2];

        for (int i = 0; i < nums.length; i++) {
            if (nbsToIndices.containsKey(nums[i])) {
                output[0] = nbsToIndices.get(nums[i]);
                output[1] = i;
                break;
            }
            nbsToIndices.put(target-nums[i], i);
        }

        return output;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-3.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m,n = len(matrix), len(matrix[0])
        i,j = 0, m*n-1

        while i <= j:
            mid = (i+j) // 2
            r,c = mid//n, mid%n

            if matrix[r][c] == target:
                return True
            
            if matrix[r][c] < target:
                i = mid+1
            else:
                j = mid-1

        return False       